bookdown/08-use-cases.Rmd

```diff
@@ -1,3 +1,405 @@
+    ParamDbl$new(id = "subsample", lower = .7, upper = .8),
+    ParamDbl$new(id = "colsample_bytree",  lower = .9, upper = 1),
+    ParamDbl$new(id = "colsample_bylevel", lower = .5, upper = .7),
+    ParamInt$new(id = "nrounds", lower = 1L, upper = 25)
+))
+
+# Define the Terminator
+terminator = TerminatorEvaluations$new(10)
+at = AutoTuner$new(lrn_xgb, "cv3", measures = "regr.mse", ps,
+  terminator, tuner = TunerRandomSearch, tuner_settings = list())
+```
+
+```{r, echo = FALSE, results = 'hide'}
+# And resample the AutoTuner
+res = resample(tsk_nozip$clone()$filter(train.idx), at, cv3)
+```
+
+```{r}
+res$performance("regr.mse")
+sprintf("RMSE of the tuned xgboost: %s", round(sqrt(res$aggregate()), 2))
+```
+
+We can obtain the resulting params in the respective splits by accessing the `r ref("ResampleResult")`.
+
+```{r}
+sapply(res$learners, function(x) x$param_set$values)
+```
+
+**NOTE:** In order to get fast results, we decide to tune only parts of the hyperparameter space that can be adjusted for `r cran_pkg("xgboost")`.
+Additionally, we only allow for $10$ iterations of random search, which is usually to little for real-world applications.
+Nonetheless, we are able to obtain an improved performance when comparing to the `r cran_pkg("ranger")` model.
+
+In order to further improve our results we have several options:
+
+- Find or Engineer better features
+- Remove Features to avoid overfitting
+- Obtain additional data (often prohbitive)
+- Try more models
+- Improve the tuning
+   - Increase the tuning budget
+   - Enlarge the tuning search space
+   - Use a more efficient tuning algorithm
+- Stacking and Ensembleing (See Pipelines Chapter!)
+
+Below we will investigate some of those possibilities and investigate whether this improves performance.
+
+### Engineer better Features: Deal with ZIP-Codes
+
+In order to better cluster the zip codes, we compute a new feature: **med_price**:
+It computes the median price in each zip-code.
+This might help our model to improve the prediction.
+
+```{r}
+# Create a new feature and append it to the task
+zip_price = house_sales_prediction[, .(med_price = median(price)),by = zipcode]
+
+# Join on the original data to match with original columns, then cbind to the task
+tsk$cbind(house_sales_prediction[zip_price, on="zipcode"][,"med_price"])
+```
+
+Again, we run `resample` and compute the **RMSE**.
+
+```{r}
+lrn_ranger = mlr_learners$get("regr.ranger")
+res = resample(task = tsk$clone()$filter(train.idx), lrn_ranger, cv3)
+res$performance("regr.mse")
+sprintf("RMSE of ranger with med_price: %s", round(sqrt(res$aggregate()), 2))
+```
+
+### Obtaining a sparser model
+
+In many cases, we might want to have a sparse model.
+For this purpose we can use a `r ref("mlr3featse::Filter")` implemented in `r mlr_pkg("mlr3featsel")`.
+This can prevent our learner from overfitting, and at the same time makes it easier for humans to interpret models, as fewer variables influence the resulting prediction.
```

Comment:
```suggestion
This can prevent our learner from overfitting make it easier for humans to interpret models as fewer variables influence the resulting prediction.
```